Add Best Practices for a Secure Supply Chain (#2309)
NuGet/docs.microsoft.com-nuget: docs/concepts/Security-Best-Practices.md

---

> title: Best Practices for a Secure Supply Chain

Reviewer: ```suggestion
title: Best practices for a secure supply chain
```

description: Best practices for securing your supply chain software using NuGet & GitHub.
author: JonDouglas
ms.author: jodou
ms.date: 02/08/2021
ms.topic: conceptual
---

> # Best Practices for a Secure Supply Chain

Reviewer: ```suggestion
# Best practices for a secure supply chain
```

Reviewer: Looks like NuGet was converted to sentence caps for heading a while ago; that is the official standard these days.

> Open Source is everywhere. It is in many proprietary codebases and community projects. For organizations and individuals, the question today is not that you are or are not using open-source code, but what open-source code you are using, and how much.

Reviewer: ```suggestion
Open Source is everywhere. It is in many proprietary codebases and community projects. For organizations and individuals, the question today is not whether you are or are not using open-source code, but what open-source code you are using, and how much.
```

> If you are not aware of what is in your software supply chain, an upstream vulnerability in one of your dependencies can be fatal, making you, and your customers, vulnerable to a potential compromise. In this document, we will dive deeper into what the term “software supply chain” means, why it matters, and how you can help secure your project’s supply chain with best practices.

Reviewer: ```suggestion
If you're not aware of what is in your software supply chain, an upstream vulnerability in one of your dependencies can be fatal, making you, and your customers, vulnerable to a potential compromise. In this document, we will dive deeper into what the term “software supply chain” means, why it matters, and how you can help secure your project’s supply chain with best practices.
```

> ![The State of the Octoverse 2020 - Open Source](media/opensource-percent.png)

Reviewer 2: Should we mention source for this any other charts we share on this doc? /cc: @Mikejo5000

Author: Each image has an alt source & there's a link to the report at the bottom. It should be fine.

## Dependencies 

> The term software supply chain is used to refer to everything that goes into your software and where it comes from. It is the dependencies and properties of your dependencies that your software supply chain depends on. A dependency is what your software needs to run. It can be code, binaries, or other components, and where they come from such as a repository or package manager.

Reviewer: ```suggestion
The term software supply chain is used to refer to everything that goes into your software and where it comes from. It is the dependencies and properties of your dependencies that your software supply chain depends on. A dependency is what your software needs to run. It can be code, binaries, or other components, and where they come from, such as a repository or package manager.
```

It includes who wrote the code, when it was contributed, how it was reviewed for security issues, known vulnerabilities, supported versions, license information, and just about anything that touches it at any point of the process.

> Your supply chain also encompasses other parts of your stack beyond a single application such as your build and packaging scripts or the software that runs the infrastructure your application relies on.

Reviewer: ```suggestion
Your supply chain also encompasses other parts of your stack beyond a single application, such as your build and packaging scripts or the software that runs the infrastructure your application relies on.
```

## Vulnerabilities

Today, software dependencies are pervasive. It is quite common for your projects to use hundreds of open-source dependencies for functionality that you did not have to write yourself. This may mean that most of your application consists of code that you did not author. 

![The State of the Octoverse 2020 - Dependencies](media/dependencies.png)

Possible vulnerabilities in your third-party or open-source dependencies, are presumably dependencies you cannot control as tightly as the code you write, which can create potential security risks in your supply chain.

If one of these dependencies has a vulnerability, the chances are you have a vulnerability as well. This can be scary as one of your dependencies may change without you even knowing. Even if a vulnerability exists in a dependency today, but is not exploitable, it can be exploitable in the future. 

> Being able to leverage the work of thousands of open-source developers & library authors means that thousands of strangers can effectively contribute directly to your production code. Your product, through your software supply chain, is affected by unpatched vulnerabilities, innocent mistakes, or even malicious attacks against dependencies.

Reviewer: ```suggestion
Being able to leverage the work of thousands of open-source developers and library authors means that thousands of strangers can effectively contribute directly to your production code. Your product, through your software supply chain, is affected by unpatched vulnerabilities, innocent mistakes, or even malicious attacks against dependencies.
```

> ## Supply Chain Compromises

Reviewer: ```suggestion
## Supply chain compromises
```

> The traditional definition of a supply chain comes from manufacturing, it is the chain of processes required to make and supply something. It includes planning, supply of materials, manufacturing, and retail. A software supply chain is similar, except instead of materials, it is code. Instead of manufacturing it is development. Instead of digging ore from the ground, code is sourced from suppliers, commercial or open source, and, in general, the open-source code comes from repositories. Adding code from a repository means your product takes a dependency on that code.

Reviewer: ```suggestion
The traditional definition of a supply chain comes from manufacturing; it is the chain of processes required to make and supply something. It includes planning, supply of materials, manufacturing, and retail. A software supply chain is similar, except instead of materials, it is code. Instead of manufacturing, it is development. Instead of digging ore from the ground, code is sourced from suppliers, commercial or open source, and, in general, the open-source code comes from repositories. Adding code from a repository means your product takes a dependency on that code.
```

One example of a software supply chain attack occurs when malicious code is purposefully added to a dependency, using the supply chain of that dependency to distribute the code to its victims. Supply chain attacks are real. There are many methods to attack a supply chain, from directly inserting malicious code as a new contributor, to taking over a contributor’s account without others noticing, or even compromising a signing key to distribute software that is not officially part of the dependency.

> Software supply chain attacks are in and of itself rarely the end goal, rather it is the beginning of an opportunity for an attacker to insert malware or provide a backdoor for future access.

Reviewer: ```suggestion
A software supply chain attack is in and of itself rarely the end goal, rather it is the beginning of an opportunity for an attacker to insert malware or provide a backdoor for future access.
```

![The State of the Octoverse 2020 - Vulnerability Lifecycle](media/vulnerability-lifecycle.png)

> ## Unpatched Software

Reviewer: ```suggestion
## Unpatched software
```

The use of open source today is significant and is not expected to slow down anytime soon. Given that we are not going to stop using open-source software, the threat to supply chain security is unpatched software. Knowing that, how can you address the risk that a dependency of your project has a vulnerability?

- **Knowing what is in your environment.** This requires discovering your dependencies and any transitive dependencies to understand the risks of those dependencies such as vulnerabilities or licensing restrictions.
- **Manage your dependencies.** When a new security vulnerability is discovered, you must determine whether you are impacted, and if so, update to the latest version and security patch available. This is especially important to review changes that introduce new dependencies or regularly auditing older dependencies.

> - **Monitor your supply chain.** This is by auditing the controls you have in place to manage your dependencies. This will help you enforce more restrictive conditions to be met for your dependencies.

Reviewer 2: Find this a bit vague and somehow this does not connect with the topics discussed later in the post. It is more about "Publishing securely"?

Author: The controls you have as an author vs a consumer differ. Authors have many more controls than consumers for monitoring supply chain.

Reviewer 3: so perhaps rephrase as "As a package publisher, take advantage of the controls you have to ensure secure supply chain for consumers of your packages"

> - **Monitor your supply chain.** This is by auditing the controls you have in place to manage your dependencies. This will help you enforce more restrictive conditions to be met for your dependencies.

Reviewer 3: 'This is by auditing the controls you have in place to manage your dependencies. This will help you enforce more restrictive conditions to be met for your dependencies.' 

I don't understand what "auditing the controls you have in place" means. Can you explain/rephrase?

Author: Meaning that these are things within your control as an author.

![The State of the Octoverse 2020 - Advisories](media/advisories.png)

> We will cover various tools & techniques that NuGet & GitHub provides that you can use today to address potential risks inside your project. 

Reviewer: ```suggestion
We will cover various tools and techniques that NuGet and GitHub provides, which you can use today to address potential risks inside your project. 
```

## Knowing what is in your environment

> ### NuGet Dependency Graph

Reviewer 3: Listing the packages in my graph doesn't provide information about their vulnerabilities.. what is the call for action here?  What should I do after I list them? IMHO this should be replaced by using tooling to indentify deprecated/vulnerable packages.

Author: I know this sounds very basic, but just knowing what dependencies you have & their versions is a best practice. For example knowing you don't use a package can help you remove it. Seeing a typo in a package might lead to a wrong dependency, etc.

> ### NuGet Dependency Graph

Reviewer: ```suggestion
### NuGet dependency graph
```

You can view your NuGet dependencies in your project by looking directly at the respective project file.

This is typically found in one of two places:

> -	[`packages.config`](https://docs.microsoft.com/nuget/reference/packages-config) – Located in the project root.

Reviewer: ```suggestion
-	[`packages.config`](../reference/packages-config.md) – Located in the project root.
```

> -	[`<PackageReference>`](https://docs.microsoft.com/nuget/consume-packages/package-references-in-project-files) – Located in the project file. 

Reviewer: ```suggestion
-	[`<PackageReference>`](../consume-packages/package-references-in-project-files.md) – Located in the project file. 
```

> Depending on what method you use to manage your NuGet dependencies, you can also use Visual Studio to view your dependencies directly in the [Solution Explorer](https://docs.microsoft.com/visualstudio/ide/solutions-and-projects-in-visual-studio?view=vs-2019#solution-explorer) or [NuGet Package Manager](https://docs.microsoft.com/nuget/consume-packages/install-use-packages-visual-studio).

Reviewer: ```suggestion
Depending on what method you use to manage your NuGet dependencies, you can also use Visual Studio to view your dependencies directly in [Solution Explorer](https://docs.microsoft.com/visualstudio/ide/solutions-and-projects-in-visual-studio?view=vs-2019#solution-explorer) or [NuGet Package Manager](https://docs.microsoft.com/nuget/consume-packages/install-use-packages-visual-studio).
```

Reviewer: Guidelines are to use relative links for docs in the same repo (.md required in the URL path), and top-level links are preferred for docs in another repo (/visualstudio/...). For the latter, no .md suffix is necessary.

Reviewer: One reason is to ensure that the locale is not present in the URL.

Author: I usually remove the locale after warnings but will definitely keep this in mind!

> For CLI environments, you can use the [`dotnet list package`](https://docs.microsoft.com/dotnet/core/tools/dotnet-list-package) command to list out your project or solution’s dependencies. 

Reviewer: ```suggestion
For CLI environments, you can use the [`dotnet list package`](/dotnet/core/tools/dotnet-list-package) command to list out your project or solution’s dependencies. 
```

> For more information on managing NuGet dependencies, [see the following documentation](https://docs.microsoft.com/nuget/consume-packages/overview-and-workflow).

Reviewer: ```suggestion
For more information on managing NuGet dependencies, [see the following documentation](../consume-packages/overview-and-workflow.md).
```

> ### GitHub Dependency Graph 

Reviewer 3: Consider having a section for projects hosted on GitHub and the security benefits if provides.

Author: Kay linked me to https://github.com/ossf/wg-best-practices-os-developers which might have some items to explore.

Author: Do you know of any links regarding what GitHub provides uniquely for hosting? I mostly just included security features sprinkled through this doc.

> ### GitHub Dependency Graph 

Reviewer: ```suggestion
### GitHub dependency graph 
```

You can use GitHub’s dependency graph to see the packages your project depends on and the repositories that depend on it. This can help you see any vulnerabilities detected in its dependencies.

For more information on GitHub repository dependencies, [see the following documentation](https://github.co/dependency-graph).

> ### Dependency Versions

Reviewer: ```suggestion
### Dependency versions
```

> To ensure a secure supply chain of dependencies, you will want to ensure that all your dependencies are regularly updated to the latest stable version. Your dependencies can include code you depend on, binaries you consume, tooling you use, and other components. This can include:

Reviewer 3: Consider adding: "Newer versions of your dependencies will often include patches of security vulnerabilities. Updating your dependencies to latest stable versions regularly will ensure you get those pathces."

Author: Added

-	[Visual Studio](https://visualstudio.microsoft.com/downloads/)
-	[.NET SDK & Runtime](https://dotnet.microsoft.com/download)
-	[NuGet](https://www.nuget.org/downloads)

> -	[NuGet Packages](https://docs.microsoft.com/nuget/consume-packages/reinstalling-and-updating-packages)

Reviewer: ```suggestion
-	[NuGet Packages](../consume-packages/reinstalling-and-updating-packages.md)
```

## Manage your dependencies

> ### NuGet Deprecated & Vulnerable Dependencies

Reviewer: ```suggestion
### NuGet deprecated and vulnerable dependencies
```

> You can use the [dotnet CLI](https://docs.microsoft.com/dotnet/core/tools/dotnet-list-package) to list any known deprecated or vulnerable dependencies you may have inside your project or solution. You can use the command `dotnet list package --deprecated` or `dotnet list package --vulnerable` to provide you a list of any known deprecations or vulnerabilities.

Reviewer: ```suggestion
You can use the [dotnet CLI](/dotnet/core/tools/dotnet-list-package) to list any known deprecated or vulnerable dependencies you may have inside your project or solution. You can use the command `dotnet list package --deprecated` or `dotnet list package --vulnerable` to provide you a list of any known deprecations or vulnerabilities.
```

> ### GitHub Vulnerable Dependencies

Reviewer: ```suggestion
### GitHub vulnerable dependencies
```

If your project is hosted on GitHub, you can leverage [GitHub Security](https://docs.github.com/en/free-pro-team@latest/github/finding-security-vulnerabilities-and-errors-in-your-code/automatically-scanning-your-code-for-vulnerabilities-and-errors) to find security vulnerabilities and errors in your project and Dependabot will fix them by opening up a pull request against your codebase. 

Catching vulnerable dependencies before they are introduced is one goal of the [“Shift Left”](https://en.wikipedia.org/wiki/Shift-left_testing) movement. Being able to have information about your dependencies such as their license, transitive dependencies, and the age of dependencies helps you do just that.

For more information about Dependabot alerts & security updates, [see the following documentation](https://docs.github.com/en/github/managing-security-vulnerabilities/about-alerts-for-vulnerable-dependencies).

> ### NuGet Feeds

Reviewer: ```suggestion
### NuGet feeds
```

> Packages can come from different feeds. To ensure you are secure, knowing what feed your packages are coming from is a best practice. One such best practice is the use of a single feed. You can accomplish this by using multiple upstream source feeds to bring your packages into a single feed.

Reviewer 3: More explanation is needed here on why multiple feeds is a problem. Also, using AzureDevOps is not something most our customers do, so this can't be the only recommendation we provide. 
Lock files and Client Trust policies help mitigate this (even if not 100%). We should frame it accordingly.

Reviewer 3: @JonDouglas , ping on this comment.

"When using multiple NuGet feeds, a package can be downloaded from any of the feeds. To ensure your build is predictable, we recommend using a single feed or a private feed with upstreaming capabilities. For further information see [link to whitepaper]"

Reviewer 3: Also, a recommendation from the whitepaper that doesn't appear here is to ensure packages on private feeds have a namespace, and that this namespace is reserved on nuget.org

Author: I'll reword this section now that we have a whitepaper out. Thanks for the suggestion!

Author: Added a commit

> For more information about single NuGet feeds, [see the following documentation](https://docs.microsoft.com/azure/devops/artifacts/concepts/upstream-sources?view=azure-devops).

Reviewer: ```suggestion
For more information about single NuGet feeds, [see the following documentation](/azure/devops/artifacts/concepts/upstream-sources?view=azure-devops).
```

> ### Client Trust Policies

Reviewer: ```suggestion
### Client trust policies
```

> There are policies that you can opt-into in which you require the packages you use to be signed. This allows you to trust a package author so long as it is author signed or trust a package if it is owned by a specific user or account that is repository signed by NuGet.org.

Reviewer: ```suggestion
There are policies that you can opt-into in which you require the packages you use to be signed. This allows you to trust a package author, as long as it is author signed, or trust a package if it is owned by a specific user or account that is repository signed by NuGet.org.
```

> To configure client trust policies, [see the following documentation](https://docs.microsoft.com/nuget/consume-packages/installing-signed-packages).

Reviewer: ```suggestion
To configure client trust policies, [see the following documentation](../consume-packages/installing-signed-packages.md).
```

> ### Lock Files

Reviewer: ```suggestion
### Lock files
```

Lock files store the hash of your package’s content. If the content hash of a package you want to install matches with the lock file, it will ensure package repeatability.

To enable lock files, [see the following documentation](../consume-packages/package-references-in-project-files#locking-dependencies).

> ## Monitor your supply chain

Reviewer 2: See comment above. Feel this topic doesn't connect with what's discussed below.

Author: Care to elaborate on which topic?

Knowing -> Seeing your dependency graph & versions
Managing -> Acting in some way from a consumer
Monitoring -> Acting in some way as an author

If there are better homes for these topics, let me know.

> ## Monitor your supply chain

Reviewer 3: Those are all recommendations to publishers.. the title should be appropriate. Something like "Be a secure publisher".

Author: Maybe this belongs in a separate doc? I think of monitoring mostly as an author based activity. You monitor if people squat on your package id, steal your api secrets, try to log into your accounts, reserve your prefix, see a new vulnerability reported on your package, etc.

From a consumer standpoint there's less control you have because you don't do anything but managing your dependencies from what you "know".

I could put some type of indicators or something that says "Package Consumer" | "Package Author" which might help give this more clarity as the doc expands more advice.

Reviewer 3: Agree. We need a clear separation between consumer recommendations and publishers. Both your suggestions - separate doc or just separate sections will work.

Author: Let me see what I can do. Would adding tags the sections under the title seem okay with the same structure given we will be adding as we learn more?

Author: For example:

## NuGet dependency graph
"Package Consumers" | "Package Authors"
Blah blah blah advice

Reviewer 3: I don't have a strong opinion here :) tags LGTM

> Lock files store the hash of your package’s content. If the content hash of a package you want to install matches with the lock file, it will ensure you package repeatability.

Reviewer: ```suggestion
Lock files store the hash of your package’s content. If the content hash of a package you want to install matches with the lock file, it will ensure package repeatability.
```

> To enable lock files, [see the following documentation](https://docs.microsoft.com/nuget/consume-packages/package-references-in-project-files#locking-dependencies).

Reviewer: ```suggestion
To enable lock files, [see the following documentation](../consume-packages/package-references-in-project-files#locking-dependencies.md).
```

### Publish to NuGet.org

> NuGet.org serves as a central repository to over 200,000 unique packages. Whenever you publish a package, NuGet.org will go through numerous validations and indexing that can benefit you in the long term. These can include scanning the package for viruses, [providing a repository signature](../reference/signed-packages-reference.md) on the package, and even protecting your package ID so only you can push updates to it.

Reviewer 2: Don't think publishing to nuget.org is based on the benefits provided by nuget.org. Its a choice of mostly whether its internal or external packages. 

We should cover prefix reservation as something folks should do when they are publishing publicly to nuget.org or even for private packages to prevent from substitution attacks.

Author: Is this not a best practice? https://youtu.be/UTPct8FKu8I?t=276

If anything reserving your package ID on a public registry is a best practice for that very case you mention even if the user has no intentions of publishing public package as prefix ID reservations aren't even well known until recently.

Reviewer 3: We should recommend reserving a package prefix of internal packages on nuget.org, but it differs from actually publishing your private packages on nuget.org.

Author: Remove then? I have no strong feelings.

Author: Removed in another commit

> NuGet.org serves as a central repository to over 200,000 unique packages. Whenever you publish a package, NuGet.org will go through numerous validations and indexing that can benefit you in the long term. These can include scanning the package for viruses, [providing a repository signature](../reference/signed-packages-reference.md) on the package, and even protecting your package ID so only you can push updates to it.

Reviewer 3: Again, I disagree with this recommendation. Publishing to nuget.org isn't a security benefit.

Author: Commented above with where this came from & why.

> NuGet.org serves as a central repository to over 200,000 unique packages. Whenever you publish a package, NuGet.org will go through numerous validations and indexing that can benefit you in the long term. These can include scanning the package for viruses, [providing a repository signature](https://docs.microsoft.com/nuget/reference/signed-packages-reference) on the package, and even protecting your package ID so only you can push updates to it.

Reviewer: ```suggestion
NuGet.org serves as a central repository to over 200,000 unique packages. Whenever you publish a package, NuGet.org will go through numerous validations and indexing that can benefit you in the long term. These can include scanning the package for viruses, [providing a repository signature](../reference/signed-packages-reference.md) on the package, and even protecting your package ID so only you can push updates to it.
```

> NuGet.org serves as a central repository to over 200,000 unique packages. Whenever you publish a package, NuGet.org will go through numerous validations and indexing that can benefit you in the long term. These can include scanning the package for viruses, [providing a repository signature](https://docs.microsoft.com/nuget/reference/signed-packages-reference) on the package, and even protecting your package ID so only you can push updates to it.

Reviewer 3: As stated here, publishing to nuget.org virus scans you package and repo-signs it. But those are really not the reason to publish something to nuget.org and make it available for everyone to download. I disagree with this recommendation.

Author: @dtivel / @joelverhagen How might we make this a better recommendation? Should we remove it?

> To learn more about the benefits of publishing on NuGet.org, [see the following documentation](https://docs.microsoft.com/nuget/nuget-org/publish-a-package#package-validation-and-indexing).

Reviewer: ```suggestion
To learn more about the benefits of publishing on NuGet.org, see [Package validation and indexing](../nuget-org/publish-a-package.md#package-validation-and-indexing).
```

> ### GitHub Secret Scanning

Reviewer: ```suggestion
### GitHub secret scanning
```

> ### GitHub Secret Scanning

Reviewer 3: IMO we should have a separate doc for publishers. "How to contribute to a secure NuGet supply chain". This is where we will list all the publisher best practices. This doc should be consumer targeted, otherwise the flow and recommendations are confusing.

> GitHub will scan repositories for NuGet API keys to prevent fraudulent uses of secrets that were accidentally committed. 

Reviewer: ```suggestion
GitHub scans repositories for NuGet API keys to prevent fraudulent uses of secrets that were accidentally committed. 
```

> To learn more about secrete scanning, [see the following documentation](https://docs.github.com/en/github/administering-a-repository/about-secret-scanning).

Reviewer: ```suggestion
To learn more about secret scanning, see [About secret scanning](https://docs.github.com/en/github/administering-a-repository/about-secret-scanning).
```

### Author Package Signing

> [Author signing](../reference/signed-packages-reference.md) allows a package author to stamp their identity on a package and for a consumer to verify it came from you. This protects you against content tampering and serves as a single source of truth about the origin of the package and the package authenticity. 

Reviewer 3: Author signing provides security benefits when it's combined with Client Policies. We should state that.

Author: We can link back to https://review.docs.microsoft.com/en-us/nuget/concepts/security-best-practices?branch=pr-en-us-2309#client-trust-policies which talks about it.

Author: Added a commit.

> [Author signing](https://docs.microsoft.com/nuget/reference/signed-packages-reference) allows a package author to stamp their identity on a package and for a consumer to verify it came from you. This protects you against content tampering & serves as a single source of truth about the origin of the package and the package authenticity. 

Reviewer: ```suggestion
[Author signing](../reference/signed-packages-reference.md) allows a package author to stamp their identity on a package and for a consumer to verify it came from you. This protects you against content tampering and serves as a single source of truth about the origin of the package and the package authenticity. 
```

> To author sign a package, [see the following documentation](https://docs.microsoft.com/nuget/create-packages/sign-a-package).

Reviewer: ```suggestion
To author sign a package, see [Sign a package](../create-packages/sign-a-package.md).
```

### Two-Factor Authentication (2FA)

Enabling two-factor authentication (2FA) can add an extra layer of security when [logging into your GitHub account](https://docs.github.com/en/github/authenticating-to-github/securing-your-account-with-two-factor-authentication-2fa) or the [NuGet.org public package repository](../nuget-org/individual-accounts.md#enable-two-factor-authentication-2fa). It is recommended that you enable two-factor authentication to protect your account.

> ### Package ID prefix reservation 

Reviewer 2: We need to have more here on rationale behind why folks should consider reserving prefixes when they publish their packages to nuget.org as well as when their packages are internal i.e. they should namespace their internal packages and reserve them on nuget.org to prevent substitution.

Author: Perhaps namespaces should be added as it's own section here? 

The reason there is no detailed rationale here is that the primary doc should serve as the introduction to the benefits and challenges. This doc is merely telling you that you should consider it & read more about it.

Joel recently updated https://docs.microsoft.com/en-us/nuget/nuget-org/id-prefix-reservation to clear up some things too FWIW.

> Enabling two-factor authentication (2FA) can add an extra layer of security when [logging into your GitHub account](https://docs.github.com/en/github/authenticating-to-github/securing-your-account-with-two-factor-authentication-2fa) or the [NuGet.org public package repository](https://docs.microsoft.com/nuget/nuget-org/individual-accounts#enable-two-factor-authentication-2fa). It is recommended to enable two-factor authentication to protect your account.

Reviewer: ```suggestion
Enabling two-factor authentication (2FA) can add an extra layer of security when [logging into your GitHub account](https://docs.github.com/en/github/authenticating-to-github/securing-your-account-with-two-factor-authentication-2fa) or the [NuGet.org public package repository](../nuget-org/individual-accounts.md#enable-two-factor-authentication-2fa). It is recommended that you enable two-factor authentication to protect your account.
```

> ### Package ID Prefix Reservation 

Reviewer: ```suggestion
### Package ID prefix reservation 
```

> To protect the identity of your packages, you can reserve a package ID prefix to associate a matching owner if your package ID prefix properly falls [under the following criteria](https://docs.microsoft.com/nuget/nuget-org/id-prefix-reservation#id-prefix-reservation-criteria).

Reviewer: ```suggestion
To protect the identity of your packages, you can reserve a package ID prefix to associate a matching owner if your package ID prefix properly falls under the [specified criteria](../nuget-org/id-prefix-reservation.md#id-prefix-reservation-criteria).
```

> To learn about reserving ID prefixes, [see the following documentation](https://docs.microsoft.com/nuget/nuget-org/id-prefix-reservation).

Reviewer: ```suggestion
To learn about reserving ID prefixes, see [Package ID prefix reservation](../nuget-org/id-prefix-reservation.md).
```

> ### Deprecating & Unlisting a Vulnerable Package

Reviewer: ```suggestion
### Deprecating and unlisting a vulnerable package
```

> To protect the .NET package ecosystem. When you are aware of a vulnerability in a package you have authored, do your best to deprecate & unlist the package so it is hidden from users searching for packages. If you are consuming a package that is deprecated & unlisted, you should avoid using the package.

Reviewer: ```suggestion
To protect the .NET package ecosystem. When you are aware of a vulnerability in a package you have authored, do your best to deprecate and unlist the package so it is hidden from users searching for packages. If you are consuming a package that is deprecated and unlisted, you should avoid using the package.
```

> To learn how to deprecate & unlist a package, see the following documentation on [deprecating](https://docs.microsoft.com/nuget/nuget-org/deprecate-packages) & [unlisting packages](https://docs.microsoft.com/nuget/nuget-org/policies/deleting-packages#unlisting-a-package).

Reviewer: ```suggestion
To learn how to deprecate and unlist a package, see the following documentation on [deprecating](../nuget-org/deprecate-packages.md) and [unlisting packages](../nuget-org/policies/deleting-packages.md#unlisting-a-package).
```

## Summary

> Your software supply chain is anything that goes into or affects your code. Even though supply chain compromises are real and growing in popularity, they are still rare – and so the most important thing you can do is protect your supply chain by **being aware of your dependencies, managing your dependencies** and **monitoring your supply chain.**

Reviewer: ```suggestion
Your software supply chain is anything that goes into or affects your code. Even though supply chain compromises are real and growing in popularity, they are still rare; so the most important thing you can do is protect your supply chain by **being aware of your dependencies, managing your dependencies** and **monitoring your supply chain.**
```

> You learned about various methods NuGet & [GitHub](https://docs.microsoft.com/learn/modules/maintain-secure-repository-github/) provides that are available to you today to be more effective in viewing, managing, and monitoring your supply chain.

Reviewer: ```suggestion
You learned about various methods NuGet and [GitHub](/learn/modules/maintain-secure-repository-github/) provides that are available to you today to be more effective in viewing, managing, and monitoring your supply chain.
```
